File: distro/basic-agent-sdk-sample-s2s/src/agent.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { TurnState, AgentApplication, TurnContext, MemoryStorage, CloudAdapter } from '@microsoft/agents-hosting';
import { ActivityTypes } from '@microsoft/agents-activity';
import { PROD_OBSERVABILITY_SCOPE } from '@microsoft/agents-a365-runtime';
import { ConfidentialClientApplication } from '@azure/msal-node';
import tokenCache, { createAgenticTokenCacheKey } from './token-cache.js';
import {
  A365_PARENT_SPAN_KEY,
  InvokeAgentScope,
  InvokeAgentScopeDetails,
  InferenceScope,
  InferenceDetails,
  InferenceOperationType,
  ExecuteToolScope,
  ToolCallDetails,
  OutputScope,
  AgentDetails,
  A365Request,
  ParentSpanRef,
} from '@microsoft/opentelemetry';

/**
 * Acquires an S2S token for a given scope using the agentic identity chain.
 * Same pattern as MsalTokenProvider.getAgenticInstanceToken but with a custom scope.
 */
async function getAgenticS2SToken(
  connection: { getAgenticApplicationToken: (tenantId: string, agentId: string) => Promise<string> },
  tenantId: string,
  agentId: string,
  scopes: string[]
): Promise<string> {
  const appToken = await connection.getAgenticApplicationToken(tenantId, agentId);
  const cca = new ConfidentialClientApplication({
    auth: {
      clientId: agentId,
      clientAssertion: appToken,
      authority: `https://login.microsoftonline.com/${tenantId}`,
    }
  });
  const result = await cca.acquireTokenByClientCredential({ scopes });
  if (!result?.accessToken) {
    throw new Error('Failed to acquire S2S token');
  }
  return result.accessToken;
}

export class A365Agent extends AgentApplication<TurnState> {
  constructor() {
    super({
      startTypingTimer: true,
      storage: new MemoryStorage(),
    });

    this.onActivity(ActivityTypes.Message, async (context: TurnContext, state: TurnState) => {
      await this.handleAgentMessageActivity(context, state);
    });
  }

  async handleAgentMessageActivity(turnContext: TurnContext, state: TurnState): Promise<void> {
    const userMessage = turnContext.activity.text?.trim() || '';
    if (!userMessage) return;

    const request: A365Request = { conversationId: turnContext.activity.conversation?.id };
    const agentDetails: AgentDetails = {
      agentId: turnContext.activity.recipient?.agenticAppId || 'basic-agent',
      agentName: 'BasicA365Agent-S2S',
      tenantId: turnContext.activity.recipient?.tenantId || 'unknown',
    };

    const invokeScope = InvokeAgentScope.start(request, {}, agentDetails);
    try {
      const spanCtx = invokeScope.getSpanContext();
      turnContext.turnState.set(A365_PARENT_SPAN_KEY, {
        traceId: spanCtx.traceId, spanId: spanCtx.spanId, traceFlags: spanCtx.traceFlags,
      } as ParentSpanRef);

      await this.preloadObservabilityTokenS2S(turnContext);

      const inferenceScope = InferenceScope.start(request, {
        operationName: InferenceOperationType.CHAT, model: 'gpt-4o', providerName: 'openai',
        inputTokens: 50, outputTokens: 120, finishReasons: ['stop'],
      }, agentDetails);
      await inferenceScope.withActiveSpanAsync(async () => {
        inferenceScope.recordInputMessages([userMessage]);
        inferenceScope.recordOutputMessages(['Simulated LLM response for: ' + userMessage]);
        inferenceScope.recordInputTokens(50);
        inferenceScope.recordOutputTokens(120);
        inferenceScope.recordFinishReasons(['stop']);
      });
      inferenceScope.dispose();

      const toolScope = ExecuteToolScope.start(request, {
        toolName: 'get_weather', arguments: JSON.stringify({ city: 'Seattle' }),
        toolCallId: 'call_test_001', description: 'Get the current weather for a city', toolType: 'function',
      }, agentDetails);
      await toolScope.withActiveSpanAsync(async () => {
        toolScope.recordResponse(JSON.stringify({ weather: 'sunny', temperature: '22°C' }));
      });
      toolScope.dispose();

      const responseText = 'The weather in Seattle is sunny, 22°C.';
      const outputScope = OutputScope.start(request, { messages: [responseText] }, agentDetails);
      outputScope.recordOutputMessages([responseText]);
      outputScope.dispose();

      await turnContext.sendActivity(responseText);
    } catch (error) {
      invokeScope.recordError(error instanceof Error ? error : new Error(String(error)));
      console.error('Agent invocation error:', error);
      await turnContext.sendActivity(`Error: ${error instanceof Error ? error.message : String(error)}`);
    } finally {
      invokeScope.dispose();
    }
  }

  /**
   * S2S observability token via the agentic identity chain.
   * Uses getAgenticS2SToken — same pattern as getAgenticInstanceToken but with the observability scope.
   */
  private async preloadObservabilityTokenS2S(turnContext: TurnContext): Promise<void> {
    const agentId = turnContext?.activity?.recipient?.agenticAppId ?? '';
    const tenantId = turnContext?.activity?.recipient?.tenantId ?? '';
    if (!agentId || !tenantId) return;

    try {
      const connection = (this.adapter as CloudAdapter).connectionManager.getConnection('service_connection');
      const token = await getAgenticS2SToken(connection, tenantId, agentId, [PROD_OBSERVABILITY_SCOPE]);

      const cacheKey = createAgenticTokenCacheKey(agentId, tenantId);
      tokenCache.set(cacheKey, token);
      console.log(`[S2S] Observability token cached (length=${token.length})`);
    } catch (error) {
      const msg = error instanceof Error ? error.message : String(error);
      console.error(`[S2S] Error: ${msg.substring(0, 200)}`);
    }
  }
}

export const agentApplication = new A365Agent();

File: python/distro/s2s-sample/src/agent.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

"""
S2S Observability Echo Agent.

Demonstrates Server-to-Server token acquisition for the Agent365 observability
exporter using the agentic identity chain. On each incoming message the agent:

  1. Acquires an S2S observability token via get_agentic_s2s_token
     (get_agentic_application_token + MSAL acquire_token_for_client)
  2. Caches it for the OTel exporter's token resolver
  3. Creates manual InvokeAgentScope / InferenceScope / OutputScope spans
  4. Echoes the user message back

Note: AgenticUserAuthorization.get_token returns a token scoped to
5a807f24-.../.default (bot framework), not the observability resource.
The S2S endpoint requires api://9b975845-.../.default, so the manual
two-step flow is needed.
"""

import logging
from os import environ

from dotenv import load_dotenv
from microsoft_agents.activity import load_configuration_from_env
from microsoft_agents.authentication.msal import MsalConnectionManager
from microsoft_agents.hosting.aiohttp import CloudAdapter
from microsoft_agents.hosting.core import (
    AgentApplication,
    Authorization,
    MemoryStorage,
    TurnContext,
    TurnState,
)
from microsoft.opentelemetry.a365.core import (
    AgentDetails,
    BaggageBuilder,
    InferenceCallDetails,
    InferenceOperationType,
    InferenceScope,
    InvokeAgentScope,
    InvokeAgentScopeDetails,
    OutputScope,
    Request,
)
from microsoft.opentelemetry.a365.core.models.response import Response

from utils.s2s_token import get_agentic_s2s_token
from utils.token_cache import cache_s2s_token

logger = logging.getLogger(__name__)

load_dotenv()
agents_sdk_config = load_configuration_from_env(environ)

STORAGE = MemoryStorage()
CONNECTION_MANAGER = MsalConnectionManager(**agents_sdk_config)
ADAPTER = CloudAdapter(connection_manager=CONNECTION_MANAGER)
AUTHORIZATION = Authorization(STORAGE, CONNECTION_MANAGER, **agents_sdk_config)

AGENT_APP = AgentApplication[TurnState](
    storage=STORAGE, adapter=ADAPTER, authorization=AUTHORIZATION, **agents_sdk_config
)


def _create_agent_details(context: TurnContext) -> AgentDetails:
    """Build AgentDetails from the incoming activity.

    Uses SDK helpers (get_agentic_instance_id/get_agentic_tenant_id) which read
    from recipient. Falls back to from_property for emulator compatibility where
    the agentic identity is on the sender.
    """
    agent_id = (
        context.activity.get_agentic_instance_id()
        or getattr(context.activity.from_property, "agentic_app_id", None)
    )
    tenant_id = (
        context.activity.get_agentic_tenant_id()
        or (context.activity.conversation.tenant_id if context.activity.conversation else None)
        or environ.get("TENANT_ID", "default-tenant")
    )

    return AgentDetails(
        agent_id=agent_id,
        agent_name=environ.get("AGENT_NAME", "S2S Echo Agent"),
        agent_description="Echo agent with S2S observability",
        tenant_id=tenant_id,
    )


@AGENT_APP.activity("message")
async def on_message(context: TurnContext, _state: TurnState):
    """Handle incoming messages: acquire S2S token, emit OTel scopes, echo back."""
    user_message = context.activity.text
    if not user_message:
        await context.send_activity("I didn't receive any message text.")
        return

    agent_details = _create_agent_details(context)

    # --- S2S observability token acquisition (must happen BEFORE creating spans) ---
    try:
        connection = CONNECTION_MANAGER.get_connection("SERVICE_CONNECTION")
        token = await get_agentic_s2s_token(
            connection,
            tenant_id=agent_details.tenant_id,
            agent_id=agent_details.agent_id,
        )
        cache_s2s_token(agent_details.agent_id, agent_details.tenant_id, token)
    except Exception as e:
        logger.error(f"[S2S] Failed to acquire observability token: {str(e)[:200]}")

    # --- Observability scopes (manual spans) ---
    request = Request(
        content=user_message,
        session_id=context.activity.conversation.id if context.activity.conversation else None,
    )

    try:
        # Set baggage so the A365 exporter knows which agent/tenant to resolve tokens for
        with BaggageBuilder().tenant_id(agent_details.tenant_id).agent_id(agent_details.agent_id).build():
            invoke_scope = InvokeAgentScope.start(
                request=request,
                scope_details=InvokeAgentScopeDetails(),
                agent_details=agent_details,
            )

            try:
                with invoke_scope:
                    invoke_scope.record_input_messages([user_message])

                    # Simulated inference scope (no real LLM call -- just demonstrating span creation)
                    inference_scope = InferenceScope.start(
                        request=request,
                        details=InferenceCallDetails(
                            operationName=InferenceOperationType.CHAT,
                            model="echo",
                            providerName="local",
                        ),
                        agent_details=agent_details,
                    )
                    with inference_scope:
                        inference_scope.record_input_messages([user_message])
                        inference_scope.record_output_messages([user_message])

                    # Output scope
                    response_text = f"Echo: {user_message}"
                    output_scope = OutputScope.start(
                        request=request,
                        response=Response(messages=[response_text]),
                        agent_details=agent_details,
                    )
                    with output_scope:
                        output_scope.record_output_messages([response_text])

                    await context.send_activity(response_text)

            except Exception as e:
                logger.error(f"Error in on_message: {e}")
                invoke_scope.record_error(e)
                await context.send_activity("Error processing your message. Please try again.")
    except Exception as e:
        logger.error(f"Error setting up observability scope: {e}")
        # Fallback: just echo without observability
        await context.send_activity(f"Echo: {user_message}")


@AGENT_APP.error
async def on_error(context: TurnContext, error: Exception):
    logger.error("[on_turn_error] unhandled error: %s", error)
    await context.send_activity("The agent encountered an error.")
